File: rag/reranker.js
/**
 * rag/reranker.js
 * [进阶] 重排序模块
 * 调用 RERANK_MODEL（默认 BAAI/bge-reranker-v2-m3）对粗检索结果精排
 * 若 RERANK_ENABLED=false（默认），search() 中不会调用此模块
 */

/**
 * 对候选文档进行重排序
 * @param {string}   query      用户查询
 * @param {string[]} documents  候选文档片段内容数组
 * @returns {Promise<Array<{index:number, relevance_score:number}>|null>}
 *   返回按相关性降序排列的结果，或 null（禁用时）
 */
export async function rerank(query, documents) {
  if (process.env.RERANK_ENABLED !== 'true') return null;
  if (!documents || documents.length === 0) return null;

  const model = process.env.RERANK_MODEL || 'BAAI/bge-reranker-v2-m3';

  const response = await fetch(`${process.env.LLM_BASE_URL}/rerank`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${process.env.LLM_API_KEY}`
    },
    body: JSON.stringify({
      model,
      query,
      documents,
      top_n: Math.min(3, documents.length),
      return_documents: false
    })
  });

  if (!response.ok) {
    const err = await response.text();
    throw new Error(`Rerank API 错误 (${response.status}): ${err.slice(0, 200)}`);
  }

  const data = await response.json();

  // 标准格式：{ results: [{ index, relevance_score }] }
  if (data.results && Array.isArray(data.results)) {
    return data.results.sort((a, b) => b.relevance_score - a.relevance_score);
  }

  return null;
}

File: python/phases/phase-3/completed/rag/reranker.py
"""
重排序模块（进阶）—— 调用硅基流动 Rerank API
"""
import requests
import streamlit as st


def rerank(query: str, chunks: list[dict], top_n: int = 3) -> list[dict]:
    """
    使用 bge-reranker 对候选片段重新排序。

    Args:
        query:  用户查询
        chunks: 候选片段（包含 content 字段）
        top_n:  返回数量

    Returns:
        重排后的 top_n 个片段（保留原始字段，追加 rerank_score）
    """
    api_key = st.secrets["LLM_API_KEY"]
    base_url = st.secrets["LLM_BASE_URL"].rstrip("/")
    model = st.secrets.get("RERANK_MODEL", "BAAI/bge-reranker-v2-m3")

    documents = [c["content"] for c in chunks]
    payload = {
        "model": model,
        "query": query,
        "documents": documents,
        "top_n": top_n,
    }
    resp = requests.post(
        f"{base_url}/rerank",
        json=payload,
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()

    results = []
    for item in data.get("results", []):
        idx = item["index"]
        chunk = dict(chunks[idx])
        chunk["rerank_score"] = round(item["relevance_score"], 4)
        results.append(chunk)

    return results
